devices/ICub/src/ICub.cpp

```diff
@@ -0,0 +1,345 @@
+
+        pImpl->ftSensorsMap.emplace(ft6dPrefix + pImpl->sensorNames[s],
+                                    std::shared_ptr<ICubImpl::ICubForceTorque6DSensor>{ft6d});
+    }
+
+    return true;
+}
+
+// ---------------
+// Generic Methods
+// ---------------
+
+wearable::WearableName ICub::getWearableName() const
+{
+    return pImpl->wearableName;
+}
+
+wearable::WearStatus ICub::getStatus() const
+{
+    wearable::WearStatus status = wearable::WearStatus::Ok;
+
+    for (const auto& s : getAllSensors()) {
+        if (s->getSensorStatus() != sensor::SensorStatus::Ok) {
+            status = wearable::WearStatus::Error;
+        }
+    }
+
+    // Default return status is Ok
+    return status;
+}
+
+wearable::TimeStamp ICub::getTimeStamp() const
+{
+    // Stamp count should be always zero
+    return {pImpl->timeStamp.time, 0};
+}
+
+bool ICub::close()
+{
+    return true;
+}
+
+// =========================
+// IPreciselyTimed interface
+// =========================
+yarp::os::Stamp ICub::getLastInputStamp()
+{
+    // Stamp count should be always zero
+    return yarp::os::Stamp(0, pImpl->timeStamp.time);
+}
+
+// ---------------------------
+// Implemented Sensors Methods
+// ---------------------------
+
+wearable::SensorPtr<const wearable::sensor::ISensor>
+ICub::getSensor(const wearable::sensor::SensorName name) const
+{
+    wearable::VectorOfSensorPtr<const wearable::sensor::ISensor> sensors = getAllSensors();
+    for (const auto& s : sensors) {
+        if (s->getSensorName() == name) {
+            return s;
+        }
+    }
+    yWarning() << LogPrefix << "User specified name <" << name << "> not found";
+    return nullptr;
+}
+
+wearable::VectorOfSensorPtr<const wearable::sensor::ISensor>
+ICub::getSensors(const wearable::sensor::SensorType type) const
+{
+    wearable::VectorOfSensorPtr<const wearable::sensor::ISensor> outVec;
+    switch (type) {
+        case sensor::SensorType::ForceTorque6DSensor: {
+            outVec.reserve(pImpl->ftSensorsMap.size());
+            for (const auto& ft6d : pImpl->ftSensorsMap) {
+                outVec.push_back(static_cast<std::shared_ptr<sensor::ISensor>>(ft6d.second));
+            }
+
+            break;
+        }
+        default: {
+            // yWarning() << LogPrefix << "Selected sensor type (" << static_cast<int>(type)
+            //           << ") is not supported by ICub";
+            return {};
+        }
+    }
+
+    return outVec;
+}
+
+wearable::SensorPtr<const wearable::sensor::IForceTorque6DSensor>
+ICub::getForceTorque6DSensor(const wearable::sensor::SensorName name) const
+{
+    // Check if user-provided name corresponds to an available sensor
+    if (pImpl->ftSensorsMap.find(name) == pImpl->ftSensorsMap.end()) {
+        yError() << LogPrefix << "Invalid sensor name";
+        return nullptr;
+    }
+
+    // Return a shared point to the required sensor
+    return static_cast<std::shared_ptr<sensor::IForceTorque6DSensor>>(pImpl->ftSensorsMap.at(name));
```

Comment:
Since getting sensors is not done at runtime, probably is safer to use [`dynamic_pointer_cast`](http://www.cplusplus.com/reference/memory/dynamic_pointer_cast/). Furthermore, use the typedef of `SensorPtr<const wearable::sensor::IForceTorque6DSensor>` instead of `shared_ptr`. In this way, if for any reason we want to change what `SensorPtr` is, we just have to change the typedef.

Reply:
Done

Reply:
Thinking more about this, since the map always contains an object which inherits from `IForceTorque6DSensor`, probably the cast is not strictly required. Can you try to remove it and if it complains about const correctness use [`const_pointer_cast`](http://www.cplusplus.com/reference/memory/const_pointer_cast/)?

Reply:
this was done in https://github.com/robotology/wearables/commit/c0b229ef04c9a2a09b0e073af326cb381dae60d3